### PR TITLE
fix type errors, add type tests

### LINKED ISSUE #117: Type error with pyright when creating BaseExceptionGroup
### Things to check first

- [X] I have searched the existing issues and didn't find my bug already reported there

- [X] I have checked that my bug is still present in the latest release

### Exceptiongroup version

master/

### Python version

3.11.8

### What happened?

foo.py:
```python
from exceptiongroup import BaseExceptionGroup
g = BaseExceptionGroup("", (KeyboardInterrupt(),))
```
```
$ pyright --pythonversion=3.10 foo.py
/home/h/Git/exceptiongroup/foo.py
  /home/h/Git/exceptiongroup/foo.py:2:5 - error: 
    Could not bind method "__new__" because "type[BaseExceptionGroup[_BaseExceptionT_co@BaseExceptionGroup]]" is not assignable to parameter "cls"
      Type "type[BaseExceptionGroup[_BaseExceptionT_co@BaseExceptionGroup]]" cannot be assigned to type "BaseExceptionGroup[Unknown]"
        "type[type]" is incompatible with "type[BaseExceptionGroup[Unknown]]" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

Using `ExceptionGroup` and e.g. `ValueError` gives no error.

### How can we reproduce the bug?

pyright==1.1.358

### PR DESCRIPTION
**Pull request by jakkdl**
* Fixes #117 
* Adds `tests/type_test.py` as type test, which quickly alerted me to two additional errors I also fixed:
* The cause of #117 wasn't isolated to `BaseExceptionGroup` and `ExceptionGroup` was also lacking `type[...]`, but instead of getting an error at creation it instead returning `ExceptionGroup[Unknown]`.
  * Both of these were introduced by me in #101
* `suppress` was typed as expecting instances of `BaseException`, but README was passing a class. I went with assuming the README was correct and it should only accept exception classes, but maybe it should accept either?
* I also added a random `cast` to `suppress.__exit__` so that file now type checks, though the code isn't type checked in CI so /shrug.

The type test could be much more thorough, and given that I almost instantly found an unrelated error with `suppress` I suspect there could be more issues hidden. You should definitely add repro's of any other issues you have, and I might check them out and see if they have easy fixes.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -90,8 +98,13 @@ extras = test
 commands = python -m pytest {posargs}
 usedevelop = true
 
-[testenv:{py37-,py38-,py39-,py310-,py311-,py312-,}pyright]
-deps = pyright
-commands = pyright --verifytypes exceptiongroup
+[testenv:{py37-,py38-,py39-,py310-,py311-,py312-,}typing]
```
Comment (agronholm) [inline pyproject.toml:101]: You need to update the `labels` option above too if you rename `pyright` to `typing`, yes?

Comment (jakkdl) [reply pyproject.toml:101]: Good catch, fixed.

**Comment by agronholm** [comment]
One more thing: could you add a changelog note too? I'll merge when that's done.

**Comment by jakkdl** [comment]
This PR mostly just fixes #101, which didn't have a changelog entry and was unreleased, so I wrote an entry for that, and another entry for `suppress()` being fixed.

Although you probably want a note for #112 as well

**Comment by jakkdl** [comment]
I don't really understand #112, but more or less copied over the commit message to the changelog.

**Comment by jakkdl** [comment]
It looks like the 3.12 failure is unrelated to this PR

**Comment by agronholm** [comment]
> It looks like the 3.12 failure is unrelated to this PR

Seems like this started within the last 24 hours. Perhaps caused by a Python update?

**Comment by jakkdl** [comment]
I think the test failures are unrelated to this PR?

> > It looks like the 3.12 failure is unrelated to this PR
> 
> Seems like this started within the last 24 hours. Perhaps caused by a Python update?

Yeah it went from 3.12.2 (https://github.com/agronholm/exceptiongroup/actions/runs/8706318205/job/23878615916#step:3:12) to 3.12.3 (https://github.com/agronholm/exceptiongroup/actions/runs/8736236104/job/23970716558#step:3:12)

**Comment by agronholm** [comment]
Thanks!

**Pull request merged**